capsules/extra/src/app_loader.rs

```
@@ -0,0 +1,470 @@
+    }
+
+    /// Copy data from the shared buffer with app and request kernel to
+    /// write the app data to flash.
+    fn write(&self, offset: usize, length: usize, processid: ProcessId) -> Result<(), ErrorCode> {
+        // Userspace sees memory that starts at address 0 even if it
+        // is offset in the physical memory.
+        match offset.checked_add(length) {
+            Some(result) => {
+                if length > self.new_app_length.get() || result > self.new_app_length.get() {
+                    // this means the app is out of bounds
+                    return Err(ErrorCode::INVAL);
+                }
+            }
+            None => {
+                return Err(ErrorCode::INVAL);
+            }
+        }
+        self.apps
+            .enter(processid, |_app, kernel_data| {
+                // Get the length of the correct allowed buffer.
+                let allow_buf_len = kernel_data
+                    .get_readonly_processbuffer(ro_allow::WRITE)
+                    .map_or(0, |read| read.len());
+
+                // Check that it exists.
+                if allow_buf_len == 0 || self.buffer.is_none() {
+                    return Err(ErrorCode::RESERVE);
+                }
+
+                // Shorten the length if the application did not give us
+                // enough bytes in the allowed buffer.
+                let active_len = cmp::min(length, allow_buf_len);
+
+                // copy data into the kernel buffer!
+                let _ = kernel_data
+                    .get_readonly_processbuffer(ro_allow::WRITE)
+                    .and_then(|write| {
+                        write.enter(|app_buffer| {
+                            self.buffer.map(|kernel_buffer| {
+                                // Check that the internal buffer and the buffer that was
+                                // allowed are long enough.
+                                let write_len = cmp::min(active_len, kernel_buffer.len());
```

Same thing here, what if the kernel buffer's smaller than the `active_len`? It seems like we're silently dropping data. Either we need to loop internally and perform a chunked copy (probably using a deferred call), indicate to userspace that only a part of the requested operation was carried out, or refuse with an error.